### DISCUSSION TITLE
pip install . as an option for source/developer builds?

### DISCUSSION THREAD

**Discussion by tcmetzger** [bokeh/bokeh#11676] [Build / CI]
As discussed with @ianthomas23 and @bryevdv in https://github.com/bokeh/bokeh/pull/11662#issuecomment-925729110 and https://github.com/bokeh/bokeh/pull/11662#issuecomment-925932236, we might want to consider using `pip install .` as an alternative to `python setup.py install`.

**Pro**: future-proof and in line with https://www.python.org/dev/peps/pep-0518
**Con**: Because of the more complex toolchain for building BokehJS, additional steps are required (i.e. have the required TS-related tools set up correctly)

@bryevdv suggests the following:

> It would be trivial to make `pip install .` just "work" even from a source checkout, all we have to do is make `--build-js` be the default. But it will fail if all the BokehJS toolchain is not installed.
> 
> That said, if there is interest to make `--build-js` be the default, I'm ok with it, as long as the error messages in case of failure are detailed and point to the dev guide, etc.

This sounds like a good plan to me. I'd be happy to help updating the docs to reflect whatever we decide to do!

**Comment by bryevdv** [comment]
There are a few options:

* just make `python setup.py` be equivalent to `python setup.py --build-js`. AFAIK that would be all that is needed to make `pip intall .` function. Note that this will cause a full clean build of BokehJS every install which takes ~30 seconds

* Secondarily we could make `--build-js` try to do an incremental build instead of a clean build. This was somewhat problematic in the long past with a differernt build systsm, but I think it might be fine nowadays. We could add a `--clean` option or something if users want to force a clean build 

* Regardless things just need adequate error messages in case the  BokehJS build toolchain is not installed or working

**Comment by bryevdv** [comment]
All of *that* said, I do want to make sure we are not encouraging downstream packagers to  build and distribute their own versions of BokehJS, which may differ slightly from the CDN versions, if they use different tools. That seem like a very bad place to be in.  Re-packagers (conda-forge, Anaconda, linux distros?) should package from the sdist, which has the BokehJS that we built bundled inside. So we might want to include a section for external package maintainers as well.
